File: src/probers.rs
//! Selection of probers

/// Get a prober for a particular VCS type.
pub fn get_prober(vcs_type: &str) -> Option<Box<dyn breezyshim::controldir::PyProber>> {
    match vcs_type {
        "bzr" => breezyshim::bazaar::RemoteBzrProber::new()
            .map(|prober| Box::new(prober) as Box<dyn breezyshim::controldir::PyProber>),
        "git" => breezyshim::git::RemoteGitProber::new()
            .map(|prober| Box::new(prober) as Box<dyn breezyshim::controldir::PyProber>),
        "hg" => breezyshim::mercurial::SmartHgProber::new()
            .map(|prober| Box::new(prober) as Box<dyn breezyshim::controldir::PyProber>),
        "svn" => breezyshim::subversion::SvnRepositoryProber::new()
            .map(|prober| Box::new(prober) as Box<dyn breezyshim::controldir::PyProber>),
        "fossil" => breezyshim::fossil::RemoteFossilProber::new()
            .map(|prober| Box::new(prober) as Box<dyn breezyshim::controldir::PyProber>),
        "darcs" => breezyshim::darcs::DarcsProber::new()
            .map(|prober| Box::new(prober) as Box<dyn breezyshim::controldir::PyProber>),
        "cvs" => breezyshim::cvs::CVSProber::new()
            .map(|prober| Box::new(prober) as Box<dyn breezyshim::controldir::PyProber>),

        _ => None,
    }
}

/// Select all probers relevant to a particular VCS type.
pub fn select_probers(vcs_type: Option<&str>) -> Vec<Box<dyn breezyshim::controldir::PyProber>> {
    if let Some(vcs_type) = vcs_type {
        if let Some(prober) = get_prober(vcs_type) {
            return vec![prober];
        }
        vec![]
    } else {
        breezyshim::controldir::all_probers()
    }
}

/// Select probers with the given VCS type as the first prober.
pub fn select_preferred_probers(
    vcs_type: Option<&str>,
) -> Vec<Box<dyn breezyshim::controldir::PyProber>> {
    let mut probers = breezyshim::controldir::all_probers();
    if let Some(vcs_type) = vcs_type {
        if let Some(prober) = get_prober(&vcs_type.to_lowercase()) {
            probers.insert(0, prober);
        }
    }
    probers
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_probers() {
        let _ = super::select_probers(None);

        let ps = super::select_probers(Some("bzr"));
        assert_eq!(ps.len(), 1);
    }

    #[test]
    fn test_preferred_probers() {
        let _ = super::select_preferred_probers(None);

        let ps = super::select_preferred_probers(Some("bzr"));
        assert!(ps.len() > 1);
    }
}

File: silver_platter/probers.py
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

from contextlib import suppress
from typing import List, Optional, Type

from breezy.bzr import RemoteBzrProber
from breezy.controldir import ControlDirFormat, Prober
from breezy.errors import UnsupportedFormatError
from breezy.git import RemoteGitProber


class UnsupportedVCSProber(Prober):
    def __init__(self, vcs_type) -> None:
        self.vcs_type = vcs_type

    def __eq__(self, other):
        return (isinstance(other, type(self))
                and other.vcs_type == self.vcs_type)

    def __call__(self):
        # The prober expects to be registered as a class.
        return self

    def priority(self, transport):
        return 200

    def probe_transport(self, transport):
        raise UnsupportedFormatError(
            "This VCS %s is not currently supported." % self.vcs_type
        )

    @classmethod
    def known_formats(klass):
        return []


prober_registry = {
    "bzr": RemoteBzrProber,
    "git": RemoteGitProber,
}

try:
    from breezy.plugins.fossil import RemoteFossilProber
except ImportError:
    pass
else:
    prober_registry["fossil"] = RemoteFossilProber

try:
    from breezy.plugins.svn import SvnRepositoryProber
except ImportError:
    pass
else:
    prober_registry["svn"] = SvnRepositoryProber

try:
    from breezy.plugins.hg import SmartHgProber
except ImportError:
    pass
else:
    prober_registry["hg"] = SmartHgProber

try:
    from breezy.plugins.darcs import DarcsProber
except ImportError:
    pass
else:
    prober_registry["darcs"] = DarcsProber

try:
    from breezy.plugins.cvs import CVSProber
except ImportError:
    pass
else:
    prober_registry["cvs"] = CVSProber


def select_probers(vcs_type=None):
    if vcs_type is None:
        return None
    try:
        return [prober_registry[vcs_type.lower()]]
    except KeyError:
        return [UnsupportedVCSProber(vcs_type)]


def select_preferred_probers(
        vcs_type: Optional[str] = None) -> List[Type[Prober]]:
    probers = list(ControlDirFormat.all_probers())
    if vcs_type:
        with suppress(KeyError):
            probers.insert(0, prober_registry[vcs_type.lower()])
    return probers
